test/src/unit-cppapi-global-order-writes-remote.cc

```diff
@@ -0,0 +1,470 @@
+    }
+    schema.add_attribute(a1);
+    if (is_var_) {
+      schema.add_attribute(a2);
+    }
+
+    Array::create(array_uri_, schema);
+
+    Array array(ctx_, array_uri_, TILEDB_READ);
+    CHECK(array.schema().array_type() == array_type);
+    CHECK(
+        array.schema()
+            .domain()
+            .dimension(0)
+            .template domain<uint64_t>()
+            .second == total_cell_count_);
+    if (array_type == TILEDB_SPARSE) {
+      CHECK(array.schema().capacity() == extent_);
+    }
+    array.close();
+  }
+
+  void write_array() {
+    Array array(ctx_, array_uri_, TILEDB_WRITE);
+    Query query(ctx_, array);
+    query.set_layout(TILEDB_GLOBAL_ORDER);
+
+    auto is_sparse = array.schema().array_type() == TILEDB_SPARSE;
+
+    char char_data = 'a';
+    // Start at column coordinate 1
+    uint64_t cols_start = 1;
+    // Resubmit until we reach total mbs requested
+    uint64_t submit_cell_count = submit_cell_count_;
+    for (uint64_t i = 0; i < total_cell_count_; i += submit_cell_count_) {
+      if (i + submit_cell_count_ > total_cell_count_) {
+        submit_cell_count -= (i + submit_cell_count_) - total_cell_count_;
+      }
+
+      // Generate some data to write to the array.
+      std::vector<T> data(submit_cell_count);
+      std::iota(data.begin(), data.end(), 0);
+
+      std::vector<uint8_t> validity_buffer(submit_cell_count, 0);
+      for (size_t i = 0; i < validity_buffer.size(); i += 2) {
+        validity_buffer[i] = 1;
+      }
+
+      // Handle coords for sparse case.
+      std::vector<uint64_t> cols(submit_cell_count);
+      if (is_sparse) {
+        std::iota(cols.begin(), cols.end(), cols_start);
+        query.set_data_buffer("cols", cols);
+      }
+
+      // Fixed sized attribute.
+      query.set_data_buffer("a", data);
+      if (is_nullable_) {
+        query.set_validity_buffer("a", validity_buffer);
+      }
+
+      // Variable sized attribute.
+      std::string var_data(submit_cell_count * sizeof(T), char_data++);
+      char_data = char_data > 'z' ? 'a' : char_data;
+      std::vector<uint64_t> var_offsets;
+      if (is_var_) {
+        // Generate offsets for variable sized attribute.
+        uint64_t max_step = var_data.size() / submit_cell_count_;
+        uint64_t j = 0;  // + (i * sizeof(T));
+        var_offsets.resize(submit_cell_count, j);
+        std::generate_n(var_offsets.begin() + 1, var_offsets.size() - 1, [&]() {
+          j += max_step--;
+          max_step =
+              max_step < 1 ? var_data.size() / var_offsets.size() : max_step;
+          return j;
+        });
+
+        query.set_data_buffer("var", var_data)
+            .set_offsets_buffer("var", var_offsets);
+        if (is_nullable_) {
+          query.set_validity_buffer("var", validity_buffer);
+        }
+      }
+
+      // Submit intermediate queries up to the final submission.
+      if (i + submit_cell_count_ < total_cell_count_) {
+        query.submit();
```

Comment:
Also tracked by [SC-25221](https://app.shortcut.com/tiledb-inc/story/25221) since it seems the use of these serialization wrappers would require relocating to `Query::submit`.